--- x2x.c-old	Sun Aug  6 17:37:00 2000
+++ x2x.c	Sun Aug  6 19:15:13 2000
@@ -37,14 +37,22 @@
  * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
  * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */

 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <unistd.h>
 #include <X11/Xlib.h>
 #include <X11/Xresource.h>
 #include <X11/Xutil.h>
 #include <X11/cursorfont.h>
 #include <X11/Xatom.h> /* for selection */
+#include <X11/extensions/XTest.h>
 #include <sys/types.h> /* for select */
 #include <sys/time.h> /* for select */
 #include "format.h"
@@ -64,6 +72,7 @@
 static void    ParseCommandLine();
 static Display *OpenAndCheckDisplay();
 static Bool    CheckTestExtension();
+static int     ErrorHandler();
 static void    DoX2X();
 static void    InitDpyInfo();
 static void    DoConnect();
@@ -277,6 +286,10 @@
     if (!(pShadow->dpy = OpenAndCheckDisplay(pShadow->name)))
       exit(3);
 
+  /* set error handler, 
+     so that program does not abort on non-critcal errors */
+  XSetErrorHandler(ErrorHandler);
+
   /* run the x2x loop */
   DoX2X(fromDpy, shadows->dpy);
 
@@ -482,6 +495,22 @@
 
 } /* END CheckTestExtension */
 
+int ErrorHandler(Display *disp, XErrorEvent *event) {
+
+  int bufflen = 1024;
+  char *buff;
+  buff = malloc(bufflen);
+
+  XGetErrorText(disp,event->error_code,buff,bufflen);
+#ifdef DEBUG
+  printf("x2x:ErrHandler(): Display:`%s` \t error_code:`%i` \n",
+       XDisplayName(NULL),event->error_code);
+#endif
+  printf(" %s \n",buff);
+  free(buff); 
+  return True;
+}
+
 #define X2X_DISCONNECTED    0
 #define X2X_AWAIT_RELEASE   1
 #define X2X_CONNECTED       2
